app/src/main/java/io/homeassistant/companion/android/widgets/media_player_controls/MediaPlayerControlsWidget.kt

```diff
@@ -278,10 +338,38 @@ class MediaPlayerControlsWidget : AppWidgetProvider() {
         }
     }
 
+    private suspend fun scheduleNextMediaUpdate(context: Context, appWidgetId: Int, entityId: String) {
+        Log.d(TAG, "Media Source currently playing, scheduling next update")
+        val am = context.getSystemService(Context.ALARM_SERVICE) as AlarmManager
+        val updateMediaIntent = Intent(context, MediaPlayerControlsWidget::class.java).apply {
+            action = UPDATE_MEDIA_IMAGE
+            putExtra(AppWidgetManager.EXTRA_APPWIDGET_ID, appWidgetId)
+        }
+        val pendingUpdateMediaIntent: PendingIntent = PendingIntent.getBroadcast(
+            context, 0, updateMediaIntent,
+            PendingIntent.FLAG_UPDATE_CURRENT
+        )
+        val nextUpdate: Number
+        val mediaPlayerEntity = integrationUseCase.getEntity(entityId)
+        var mediaFinishedIn = 30
+        if (mediaPlayerEntity.attributes["media_duration"] != null && mediaPlayerEntity.attributes["media_position"] != null) {
+            mediaFinishedIn =
+                (mediaPlayerEntity.attributes["media_duration"] as Double).minus(mediaPlayerEntity.attributes["media_position"] as Double)
+                    .toInt()
+        }
+        if (mediaFinishedIn < 30) {
+            nextUpdate = mediaFinishedIn * 1000
+        } else {
+            nextUpdate = 30000
+        }
+        am.set(AlarmManager.RTC_WAKEUP, System.currentTimeMillis().plus(nextUpdate), pendingUpdateMediaIntent)
```

Comment:
Instead of using alarm manager to schedule an update (which will interfere with the next alarm sensor), we should update the widget with the websocket updates as we subscribe to entity updates and can update as soon as the media player changes its state. Take a look at the entity widget and template widget to see how those update as soon as a state change is detected. This should help simplify the code here a bit and its a new feature that the app just gained that we should continue to use because we are already subscribed to these update :)  It should also remove the need for all the additional `getEntity` calls too

Reply:
Okay, ill take a look